src/main/kotlin/fi/metatavu/lipsanen/projects/milestones/tasks/TaskController.kt

```diff
@@ -172,4 +205,201 @@ class TaskController {
+     * @param newStartDate new start date
+     * @param newEndDate new end date
+     * @param milestone milestone
+     * @param proposalMode if the task update happens in proposal mode - if the task update happens in proposal mode - in this case reject the dependent proposals that affect the tasks affected by the update
+     * @return updated task
+     * @throws TaskOutsideMilestoneException if the cascade update goes out of the milestone boundaries
+     */
+    private suspend fun updateTaskDates(
+        movableTask: TaskEntity,
+        newStartDate: LocalDate,
+        newEndDate: LocalDate,
+        milestone: MilestoneEntity,
+        proposalMode: Boolean = false
+    ): TaskEntity {
+        if (movableTask.startDate == newStartDate && movableTask.endDate == newEndDate) {
+            return movableTask
+        }
+
+        val moveForward = movableTask.endDate < newEndDate
+        val moveBackward = movableTask.startDate > newStartDate
+        movableTask.startDate = newStartDate
+        movableTask.endDate = newEndDate
+
+        taskEntityRepository.persistSuspending(movableTask) //Save the task at this point so that when listing connections it is up-to-date
+        Panache.flush()
+
+        val updatableTasks = mutableListOf<TaskEntity>()
+        if (moveForward) {
+            updateTaskConnectionsForward(movableTask, updatableTasks)
+        }
+        if (moveBackward) {
+            updateTaskConnectionsBackward(movableTask, updatableTasks)
+        }
+
+        // Check if the dependent tasks are still within the milestone
+        updatableTasks.forEach {
+            if (it.startDate < milestone.startDate || it.endDate > milestone.endDate) {
+                throw TaskOutsideMilestoneException(it.id, it.startDate, it.endDate)
+            }
+        }
+        updatableTasks.forEach { taskEntityRepository.persistSuspending(it) }
+
+        //in proposal model cancel all the proposals that affect the moved tasks
+        if (proposalMode) {
+            proposalController.list(updatableTasks.distinctBy { it.id }).forEach {
+                it.status = ChangeProposalStatus.REJECTED
+                proposalController.persist(it)
+            }
+        }
+
+        return movableTask
+    }
+
+    /**
+     * Cascade update task connections backward (all parents of movedTask recursively)
+     *
+     * @param movedTask task to update
+     * @param updatableTasks list of tasks to update (this list is updated with new tasks to be saved)
+     */
+    private suspend fun updateTaskConnectionsBackward(movedTask: TaskEntity, updatableTasks: MutableList<TaskEntity>) {
+        val tasks = ArrayDeque<TaskEntity>()
+        tasks.add(movedTask)
+        while (tasks.isNotEmpty()) {
+            val currentTask = tasks.removeFirst()
+            val connections = taskConnectionController.list(currentTask, TaskConnectionRole.TARGET)
+            for (connection in connections) {
+                var updated = 0
```

Comment:
Is there a reason for int instead of boolean?

Reply:
No, i simply forgot replacing it to boolean 😅

Reply:
Fixed